x/did/types/did.go

```diff
@@ -0,0 +1,436 @@
+		return false
+	}
+	if !doc.validVerificationRelationships(doc.KeyAgreements) {
+		return false
+	}
+	if !doc.validVerificationRelationships(doc.CapabilityInvocations) {
+		return false
+	}
+	if !doc.validVerificationRelationships(doc.CapabilityDelegations) {
+		return false
+	}
+
+	for _, service := range doc.Services {
+		if !service.Valid() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (doc DIDDocument) validVerificationRelationships(relationships []*VerificationRelationship) bool {
+	for _, relationship := range relationships {
+		if !relationship.Valid(doc.ID) {
+			return false
+		}
+		if !relationship.hasDedicatedMethod() {
+			// if the relationship isn't a dedicated verification method,
+			// the referenced verification method must be presented in the 'verificationMethod' property.
+			if _, ok := doc.VerificationMethodByID(relationship.VerificationMethodID); !ok {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+
+func (doc DIDDocument) Empty() bool {
+	return EmptyDID(doc.ID)
+}
+
+// TODO Duplicate with protobuf
+/*func (doc DIDDocument) String() string {
+	bz, _ := json.Marshal(doc)
+	return string(bz)
+}*/
+
+// GetSignBytes returns a byte array which is used to generate a signature for verifying DID ownership.
+func (doc DIDDocument) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&doc))
+}
+
+// VerificationMethodByID finds a VerificationMethod by ID.
+// If the corresponding VerificationMethod doesn't exist, it returns a false.
+func (doc DIDDocument) VerificationMethodByID(id string) (VerificationMethod, bool) {
+	for _, verificationMethod := range doc.VerificationMethods {
+		if verificationMethod.ID == id {
+			return *verificationMethod, true
+		}
+	}
+	return VerificationMethod{}, false
+}
+
+// VerificationMethodFrom finds a VerificationMethod from the slice of VerificationRelationship by its ID.
+// There are two types of VerificationRelationship. If it has a dedicated VerificationMethod, it is returned as it is.
+// If the relationship has only a ID of VerificationMethod, this function tries to find a corresponding VerificationMethod in the DIDDocument.
+func (doc DIDDocument) VerificationMethodFrom(relationships []*VerificationRelationship, id string) (VerificationMethod, bool) {
+	for _, relationship := range relationships {
+		if relationship.VerificationMethodID == id {
+			if relationship.hasDedicatedMethod() {
+				return *relationship.DedicatedVerificationMethod, true
+			} else {
+				return doc.VerificationMethodByID(id)
+			}
+		}
+	}
+
+	return VerificationMethod{}, false
+}
+
+func ValidContexts(ctxs []string) bool {
+	if len(ctxs) == 0 || ctxs[0] != ContextDIDV1 { // the 1st one must be ContextDIDV1
+		return false
+	}
+
+	set := make(map[string]struct{}, len(ctxs))
+	for _, ctx := range ctxs {
+		_, dup := set[ctx] // check the duplication
+		if dup || !ValidContext(ctx) {
+			return false
+		}
+		set[ctx] = struct{}{}
+	}
+	return true
+}
+
```

Comment:
AS-IS 
- if `@context` one value then return string.
- if `@context` multi value then return string array.

TO-BE
- Only return string array.

Because Contexts type fixed to `[]string`. Not alias.
So it is difficult to implement `MarshalJSON, UnmarshalJSON` methods.

ref : https://github.com/medibloc/panacea-core/blob/2dfceffc647db15499c715e8833c5e379c04e028/x/did/types/did.go#L159

Reply:
Come to think of it... 
The ₩@context ₩of the existing string type will be a problem. :thinking:

Reply:
Oh. I got it. Let me think.